autokeras/auto/auto_model.py

```diff
@@ -214,9 +152,107 @@ def _add_node(self, input_node):
         if input_node not in self._node_to_id:
             self._node_to_id[input_node] = len(self._node_to_id)
 
+    def fit(self,
+            x=None,
+            y=None,
+            validation_data=None,
+            **kwargs):
+        # Initialize HyperGraph model
+        x = layer_utils.format_inputs(x, 'train_x')
+        y = layer_utils.format_inputs(y, 'train_y')
+
+        # TODO: Set the shapes only if they are not provided by the user when
+        #  initiating the HyperHead or Block.
+        for x_input, input_node in zip(x, self.inputs):
+            input_node.shape = x_input.shape[1:]
+        for y_input, output_node in zip(y, self.outputs):
+            if len(y_input.shape) == 1:
+                y_input = np.reshape(y_input, y_input.shape + (1,))
+            output_node.shape = y_input.shape[1:]
+            output_node.in_hypermodels[0].output_shape = output_node.shape
+
+        # Prepare the dataset
+        if validation_data is None:
+            (x, y), (x_val, y_val) = layer_utils.split_train_to_valid(x, y)
+            validation_data = x_val, y_val
+
+        self.tuner = kerastuner.RandomSearch(
+            hypermodel=self,
+            objective='val_loss',
+            max_trials=self.trials,
+            directory=self.directory)
+
+        # TODO: allow early stop if epochs is not specified.
+        self.tuner.search(x=x,
+                          y=y,
+                          validation_data=validation_data,
+                          **kwargs)
+
+    def predict(self, x, **kwargs):
+        """Predict the output for a given testing data. """
+        return self.tuner.get_best_models(1)[0].predict(x, **kwargs)
+
+    def _get_metrics(self):
+        metrics = []
+        for metrics_list in [output_node.in_hypermodels[0].metrics for
+                             output_node in self.outputs
+                             if isinstance(output_node.in_hypermodels[0],
+                                           hyper_head.HyperHead)]:
+            metrics += metrics_list
+        return metrics
+
     def _get_loss(self):
         loss = nest.flatten([output_node.in_hypermodels[0].loss
                              for output_node in self.outputs
                              if isinstance(output_node.in_hypermodels[0],
                                            hyper_head.HyperHead)])
         return loss
+
+    def _compiled(self, hp, model):
+        # Specify hyperparameters from compile(...)
+        optimizer = hp.Choice('optimizer',
+                              ['adam',
+                               'adadelta',
+                               'sgd'])
+
+        model.compile(optimizer=optimizer,
+                      metrics=self._get_metrics(),
+                      loss=self._get_loss())
+
+        return model
+
+
+class AutoModel(GraphAutoModel):
```

Comment:
At least from the naming perspective, `GraphAutoModel`  should inherit `AutoModel`.

Reply:
I suspect we should have a a BaseAutoModel (implementer-facing) to handle abstract logic that does not rely on the graph API, and an AutoModel subclass that is public-facing and implements the graph API.